### PR TITLE
Adding 'related' link headers for entity references

### PR DESCRIPTION
**Pull request by dannylamb**
(cherry picked from commit 10ac4e444b68c4fa8db13a04289acd68276b786b)

**GitHub Issue**: Part of https://github.com/Islandora-CLAW/CLAW/issues/716

Other Relevant Links:
- ~~https://github.com/Islandora-CLAW/islandora/pull/68 needs to get merged first.~~
- https://github.com/Islandora-CLAW/islandora_collection/pull/24 adds the collection header code back in

# What does this Pull Request do?

Adds a `rel="related"` link header to GET and HEAD response for each entity referenced by an entity reference field.  The `title` of the link header is the field's label.  So for a resource that has a tiff, a lower res display copy, and belongs to a collection, I got: 

```
<http://localhost:8000/node/2>; rel="related"; title="Member Of", <http://localhost:8000/media/2>; rel="related"; title="Tiff image", <http://localhost:8000/media/3>; rel="related"; title="Web Image"
```

I've moved the collection specific link header to the islandora_collection module in another pull request.

# What's new?
* A new `View Alter` interface
* A base class for view alters that add link headers
* A view alter that adds the related link headers, exposed as a service
* Functional tests 

# How should this be tested?

Go make an Islandora Image and then hit it with a GET or HEAD request.   You should see link headers for all the entities you made though the form.  They follow the pattern of `<node_uri>; rel="related"; title="Field Label"`  

# Interested parties
@Islandora-CLAW/committers

### CODE REVIEW

**Comment by mjordan** [comment]
Long time listener, first time caller.

In branch 'related-header-cherry-pick', I did a `drush cache-rebuild`, but am not seeing the new 'related' headers, even though my Web image is a member of a collection:

```
curl -v "http://localhost:8000/node/2?_format=jsonld"
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8000 (#0)
> GET /node/2?_format=jsonld HTTP/1.1
> User-Agent: curl/7.35.0
> Host: localhost:8000
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Tue, 09 Jan 2018 05:49:34 GMT
* Server Apache/2.4.18 (Ubuntu) is not blacklisted
< Server: Apache/2.4.18 (Ubuntu)
< X-Powered-By: PHP/7.0.22-0ubuntu0.16.04.1
< Cache-Control: must-revalidate, no-cache, private
< Link: <http://localhost:8000/node/2>; rel="canonical"
< Link: <http://localhost:8000/node/2/delete>; rel="https://drupal.org/link-relations/delete-form"
< Link: <http://localhost:8000/node/2/edit>; rel="edit-form"
< Link: <http://localhost:8000/node/2/revisions>; rel="version-history"
< Link: <http://localhost:8000/node/2>; rel="https://drupal.org/link-relations/revision"
< Link: <http://localhost:8000/node?node=2>; rel="https://drupal.org/link-relations/create"
< X-Drupal-Dynamic-Cache: MISS
< X-UA-Compatible: IE=edge
< Content-language: en
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< Expires: Sun, 19 Nov 1978 05:00:00 GMT
< X-Generator: Drupal 8 (https://www.drupal.org)
< X-Drupal-Cache: MISS
< Content-Length: 1172
< Content-Type: application/ld+json
```

Am I testing this wrong?

**Comment by dannylamb** [comment]
@mjordan I'm between environments right now, but I'll test again as soon as it's ready.  Maybe something is different after my cherry pick shenanigans.

**Comment by dannylamb** [comment]
I just tried it out and it works on http://localhost:8000/node/2 but not http://localhost:8000/node/2?_format=jsonld.  It has stopped getting triggered on REST requests for some reason.  Let me dig into this.

**Comment by dannylamb** [comment]
@mjordan Ok, so I just played around with this and managed to get it working by responding to kernel events instead of doing the alter thingy, which is better anyway if you ask me.  It's stomping over the headers for REST requests, but at least it's returning them.  Once I get it properly merging with existing headers from the other route subscribers I'll update this PR.

**Comment by mjordan** [comment]
Cool, let me know when you want me to retest.

**Comment by dannylamb** [comment]
Bear with me, I've got to do some cleanup and I'd like to re-visit the tests to make sure they cover authorization.

**Comment by mjordan** [comment]
@dannylamb I see tests are passing, let me know when you want me to test.

**Comment by dannylamb** [comment]
@mjordan Ok, this should be good to test again.

**Comment by mjordan** [comment]
I'm seeing the "Member of" but not the other related links. 

* vagrant
  * `git pull` from your feature branch
  * `drush cache-rebuil` in vagrnant then on my host
* my host
    * `curl -v "http://localhost:8000/node/2?_format=jsonld"`

 gets me:

```
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8000 (#0)
> GET /node/2?_format=jsonld HTTP/1.1
> User-Agent: curl/7.35.0
> Host: localhost:8000
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Fri, 12 Jan 2018 14:52:41 GMT
* Server Apache/2.4.18 (Ubuntu) is not blacklisted
< Server: Apache/2.4.18 (Ubuntu)
< X-Powered-By: PHP/7.0.22-0ubuntu0.16.04.1
< Cache-Control: must-revalidate, no-cache, private
< Link: <http://localhost:8000/node/2>; rel="canonical"
< Link: <http://localhost:8000/node/2/delete>; rel="https://drupal.org/link-relations/delete-form"
< Link: <http://localhost:8000/node/2/edit>; rel="edit-form"
< Link: <http://localhost:8000/node/2/revisions>; rel="version-history"
< Link: <http://localhost:8000/node/2>; rel="https://drupal.org/link-relations/revision"
< Link: <http://localhost:8000/node?node=2>; rel="https://drupal.org/link-relations/create"
< Link: <http://localhost:8000/node/1>; rel="related"; title="Member Of"
< X-Drupal-Dynamic-Cache: MISS
< X-UA-Compatible: IE=edge
< Content-language: en
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< Expires: Sun, 19 Nov 1978 05:00:00 GMT
< X-Generator: Drupal 8 (https://www.drupal.org)
< X-Drupal-Cache: MISS
< Content-Length: 1172
< Content-Type: application/ld+json
< 
[response body here...]
```

**Comment by mjordan** [comment]
woops, I see travis hasn't finished yet. Let me try again when that happens. :disappointed:

**Comment by mjordan** [comment]
Same result....

**Comment by dannylamb** [comment]
@mjordan It's because you're doing the REST requests as anonymous, which can view published content but not media.  Enable basic authentication using restui and try passing your creds to curl.

**Comment by mjordan** [comment]
I've confirmed Basic is enabled for both content and media, and when I issue

```curl -v -u admin:Islandora "http://localhost:8000/node/2?_format=jsonld"```

I now get

```
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8000 (#0)
* Server auth using Basic with user 'admin'
> GET /node/2?_format=jsonld HTTP/1.1
> Authorization: Basic YWRtaW46SXNsYW5kb3Jh
> User-Agent: curl/7.35.0
> Host: localhost:8000
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Fri, 12 Jan 2018 16:06:29 GMT
* Server Apache/2.4.18 (Ubuntu) is not blacklisted
< Server: Apache/2.4.18 (Ubuntu)
< X-Powered-By: PHP/7.0.22-0ubuntu0.16.04.1
< Cache-Control: must-revalidate, no-cache, private
< Link: <http://localhost:8000/node/2>; rel="canonical"
< Link: <http://localhost:8000/node/2/delete>; rel="https://drupal.org/link-relations/delete-form"
< Link: <http://localhost:8000/node/2/edit>; rel="edit-form"
< Link: <http://localhost:8000/node/2/revisions>; rel="version-history"
< Link: <http://localhost:8000/node/2>; rel="https://drupal.org/link-relations/revision"
< Link: <http://localhost:8000/node?node=2>; rel="https://drupal.org/link-relations/create"
< Link: <http://localhost:8000/node/1>; rel="related"; title="Member Of"
< Link: <http://localhost:8000/node/1>; rel="related"; title="Member Of"
< X-Drupal-Dynamic-Cache: HIT
< X-UA-Compatible: IE=edge
< Content-language: en
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< Expires: Sun, 19 Nov 1978 05:00:00 GMT
< X-Generator: Drupal 8 (https://www.drupal.org)
< Content-Length: 1172
< Content-Type: application/ld+json
< 
```

Note that "Member of" is showing up twice now.

**Comment by mjordan** [comment]
@dannylamb `drush cache-rebuild all` worked, sorta.... now I'm not getting dupes but I'm also not getting the binary type property:

```
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8000 (#0)
* Server auth using Basic with user 'admin'
> GET /node/2?_format=jsonld HTTP/1.1
> Authorization: Basic YWRtaW46SXNsYW5kb3Jh
> User-Agent: curl/7.35.0
> Host: localhost:8000
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Fri, 12 Jan 2018 18:18:58 GMT
* Server Apache/2.4.18 (Ubuntu) is not blacklisted
< Server: Apache/2.4.18 (Ubuntu)
< X-Powered-By: PHP/7.0.22-0ubuntu0.16.04.1
< Cache-Control: must-revalidate, no-cache, private
< Link: <http://localhost:8000/node/2>; rel="canonical"
< Link: <http://localhost:8000/node/2/delete>; rel="https://drupal.org/link-relations/delete-form"
< Link: <http://localhost:8000/node/2/edit>; rel="edit-form"
< Link: <http://localhost:8000/node/2/revisions>; rel="version-history"
< Link: <http://localhost:8000/node/2>; rel="https://drupal.org/link-relations/revision"
< Link: <http://localhost:8000/node?node=2>; rel="https://drupal.org/link-relations/create"
< Link: <http://localhost:8000/node/1>; rel="related"; title="Member Of"
< X-Drupal-Dynamic-Cache: MISS
< X-UA-Compatible: IE=edge
< Content-language: en
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< Expires: Sun, 19 Nov 1978 05:00:00 GMT
< X-Generator: Drupal 8 (https://www.drupal.org)
< Content-Length: 1172
< Content-Type: application/ld+json
```

In case it matters, I'm running 937901a55c5c5407ba3b61a9c10b99ce60fa7e40.

**Comment by dannylamb** [comment]
@mjordan Yeah, that's what i'm on too.  Weird...

So two things are coming to mind.  One is that I've set the priority wrong on the event listener, and it's firing too early or late (i think early?)  The other is the interplay with caching as @whikloj pointed out.  I have a feeling the view_alter somehow took care of things I have to handle myself now that using events.

**Comment by dannylamb** [comment]
@mjordan No need to retest, I've got to investigate some things.

**Comment by dannylamb** [comment]
@whikloj @mjordan Yeah, it's the cache.  On an initial request: it adds the link headers and caches them.  On a second request it pulls the response out of the cache and then adds the headers again and returns.

:man_facepalming:

**Comment by dannylamb** [comment]
Ok, so it was both a timing and a cache issue.  I dug through core to see how they do it and found out the just check the headers!

I also enabled debug headers for caching and inspected everything and it all seems ok.  Since it's all core stuff involved, they set cacheability metadata appropriately.

@mjordan @whikloj It's ready to try again, though I'm still concerned as to why @mjordan can't see the media references as admin.

**Comment by whikloj** [comment]
```
ubuntu@claw:/var/www/html/drupal/web/modules/contrib/islandora$ curl -I -u admin:islandora "http://localhost:8000/node/2?_format=jsonld" 
HTTP/1.1 200 OK
Date: Fri, 12 Jan 2018 21:50:13 GMT
Server: Apache/2.4.18 (Ubuntu)
X-Powered-By: PHP/7.0.22-0ubuntu0.16.04.1
Cache-Control: must-revalidate, no-cache, private
Link: <http://localhost:8000/node/2>; rel="canonical"
Link: <http://localhost:8000/node/2/delete>; rel="https://drupal.org/link-relations/delete-form"
Link: <http://localhost:8000/node/2/edit>; rel="edit-form"
Link: <http://localhost:8000/node/2/revisions>; rel="version-history"
Link: <http://localhost:8000/node/2>; rel="https://drupal.org/link-relations/revision"
Link: <http://localhost:8000/node?node=2>; rel="https://drupal.org/link-relations/create"
Link: <http://localhost:8000/node/1>; rel="related"; title="Member Of"
Link: <http://localhost:8000/media/1>; rel="related"; title="TN"
Link: <http://localhost:8000/media/2>; rel="related"; title="Web Image"
X-Drupal-Dynamic-Cache: HIT
X-UA-Compatible: IE=edge
Content-language: en
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Expires: Sun, 19 Nov 1978 05:00:00 GMT
X-Generator: Drupal 8 (https://www.drupal.org)
Content-Type: application/ld+json
```

This appears correct and consistent to me on multiple separate requests.

**Comment by mjordan** [comment]
Running 7a9e7e846eea65b4cb6fcd6a82d4c025ed91d037 and clearing my cache gives me the same results as before - I'm getting the "Member of" entry but not the binary labels. I also tried with a new TIFF image and tried my nodes a few times in a row. Same thing.

Could it be possible that I'm missing some other dependencies outside of Danny's branch that I'm missing, for example?

**Comment by whikloj** [comment]
@mjordan Not that I can see, I tried making a new image with a Tiff and Web Image in case it was on the Tiff and it seems to work too. 

Did you uninstall the islandora module, then pull the new commit, then re-install it? 

Drupal 8 can have some weird things, also when you are clearing the cache is that using drush, drupal-console or from the GUI?

**Comment by whikloj** [comment]
Just for completeness sake
```
ubuntu@claw:/var/www/html/drupal/web/modules/contrib/islandora$ curl -I -u admin:islandora "http://localhost:8000/node/3?_format=jsonld" 
HTTP/1.1 200 OK
Date: Fri, 12 Jan 2018 22:00:51 GMT
Server: Apache/2.4.18 (Ubuntu)
X-Powered-By: PHP/7.0.22-0ubuntu0.16.04.1
Cache-Control: must-revalidate, no-cache, private
Link: <http://localhost:8000/node/3>; rel="canonical"
Link: <http://localhost:8000/node/3/delete>; rel="https://drupal.org/link-relations/delete-form"
Link: <http://localhost:8000/node/3/edit>; rel="edit-form"
Link: <http://localhost:8000/node/3/revisions>; rel="version-history"
Link: <http://localhost:8000/node/3>; rel="https://drupal.org/link-relations/revision"
Link: <http://localhost:8000/node?node=3>; rel="https://drupal.org/link-relations/create"
Link: <http://localhost:8000/node/1>; rel="related"; title="Member Of"
Link: <http://localhost:8000/media/3>; rel="related"; title="Tiff image"
Link: <http://localhost:8000/media/4>; rel="related"; title="Web Image"
X-Drupal-Dynamic-Cache: HIT
X-UA-Compatible: IE=edge
Content-language: en
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Expires: Sun, 19 Nov 1978 05:00:00 GMT
X-Generator: Drupal 8 (https://www.drupal.org)
Content-Type: application/ld+json
```

**Comment by mjordan** [comment]
@whikloj 

> Did you uninstall the islandora module, then pull the new commit, then re-install it?

Nope, I just pulled in the changes from @dannylamb 's branch. Is that not sufficient?

>  when you are clearing the cache is that using drush, drupal-console or from the GUI?

`drush cache-rebuild all`

**Comment by whikloj** [comment]
@mjordan I would try `drupal router:rebuild` then `drupal cr all`. I'm wondering if the status code change is in the routes and was not affected by the cache clearing?

**Comment by mjordan** [comment]
Thanks, did both but no effect:

```
* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8000 (#0)
* Server auth using Basic with user 'admin'
> GET /node/2?_format=jsonld HTTP/1.1
> Authorization: Basic YWRtaW46SXNsYW5kb3Jh
> User-Agent: curl/7.35.0
> Host: localhost:8000
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Fri, 12 Jan 2018 22:14:13 GMT
* Server Apache/2.4.18 (Ubuntu) is not blacklisted
< Server: Apache/2.4.18 (Ubuntu)
< X-Powered-By: PHP/7.0.22-0ubuntu0.16.04.1
< Cache-Control: must-revalidate, no-cache, private
< Link: <http://localhost:8000/node/2>; rel="canonical"
< Link: <http://localhost:8000/node/2/delete>; rel="https://drupal.org/link-relations/delete-form"
< Link: <http://localhost:8000/node/2/edit>; rel="edit-form"
< Link: <http://localhost:8000/node/2/revisions>; rel="version-history"
< Link: <http://localhost:8000/node/2>; rel="https://drupal.org/link-relations/revision"
< Link: <http://localhost:8000/node?node=2>; rel="https://drupal.org/link-relations/create"
< Link: <http://localhost:8000/node/1>; rel="related"; title="Member Of"
< X-Drupal-Dynamic-Cache: MISS
< X-UA-Compatible: IE=edge
< Content-language: en
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< Expires: Sun, 19 Nov 1978 05:00:00 GMT
< X-Generator: Drupal 8 (https://www.drupal.org)
< Content-Length: 1172
< Content-Type: application/ld+json
```
:slightly_frowning_face:

**Comment by dannylamb** [comment]
If you're pulling in the changes on top of a fresh claw-playbook install
you should be fine.

I also tested by using composer to set up a fresh drupal-project and
require islandora_image. Pulled in my changes to the islandora module and
configured the REST endpoints and it worked fine.



On Jan 12, 2018 6:15 PM, "Mark Jordan" <notifications@github.com> wrote:

Thanks, did both but no effect:

* Hostname was NOT found in DNS cache
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8000 (#0)
* Server auth using Basic with user 'admin'
> GET /node/2?_format=jsonld HTTP/1.1
> Authorization: Basic YWRtaW46SXNsYW5kb3Jh
> User-Agent: curl/7.35.0
> Host: localhost:8000
> Accept: */*
>
< HTTP/1.1 200 OK
< Date: Fri, 12 Jan 2018 22:14:13 GMT
* Server Apache/2.4.18 (Ubuntu) is not blacklisted
< Server: Apache/2.4.18 (Ubuntu)
< X-Powered-By: PHP/7.0.22-0ubuntu0.16.04.1
< Cache-Control: must-revalidate, no-cache, private
< Link: <http://localhost:8000/node/2>; rel="canonical"
< Link: <http://localhost:8000/node/2/delete>;
rel="https://drupal.org/link-relations/delete-form"
< Link: <http://localhost:8000/node/2/edit>; rel="edit-form"
< Link: <http://localhost:8000/node/2/revisions>; rel="version-history"
< Link: <http://localhost:8000/node/2>;
rel="https://drupal.org/link-relations/revision"
< Link: <http://localhost:8000/node?node=2>;
rel="https://drupal.org/link-relations/create"
< Link: <http://localhost:8000/node/1>; rel="related"; title="Member Of"
< X-Drupal-Dynamic-Cache: MISS
< X-UA-Compatible: IE=edge
< Content-language: en
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< Expires: Sun, 19 Nov 1978 05:00:00 GMT
< X-Generator: Drupal 8 (https://www.drupal.org)
< Content-Length: 1172
< Content-Type: application/ld+json

🙁

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/Islandora-CLAW/islandora/pull/71#issuecomment-357369409>,
or mute the thread

**Comment by mjordan** [comment]
OK, I'll rebuilt my vagrant and try again later today.

**Comment by mjordan** [comment]
* Fresh claw-playbook install, checked out @dannylamb's `related-header-cherry-pick` branch.
* Running `drupal cache:rebuild all` resulted in the following response to `curl -v -u admin:Islandora "http://localhost:8000/node/2?_format=jsonld"`:

```
< HTTP/1.1 200 OK
< Date: Sat, 13 Jan 2018 21:14:59 GMT
* Server Apache/2.4.18 (Ubuntu) is not blacklisted
< Server: Apache/2.4.18 (Ubuntu)
< X-Powered-By: PHP/7.0.22-0ubuntu0.16.04.1
< Cache-Control: must-revalidate, no-cache, private
< Link: <http://localhost:8000/node/2>; rel="canonical"
< Link: <http://localhost:8000/node/2/delete>; rel="https://drupal.org/link-relations/delete-form"
< Link: <http://localhost:8000/node/2/edit>; rel="edit-form"
< Link: <http://localhost:8000/node/2/revisions>; rel="version-history"
< Link: <http://localhost:8000/node/2>; rel="https://drupal.org/link-relations/revision"
< Link: <http://localhost:8000/node?node=2>; rel="https://drupal.org/link-relations/create"
< Link: <http://localhost:8000/node/1>; rel="related"; title="Member Of"
< X-Drupal-Dynamic-Cache: MISS
< X-UA-Compatible: IE=edge
< Content-language: en
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< Expires: Sun, 19 Nov 1978 05:00:00 GMT
< X-Generator: Drupal 8 (https://www.drupal.org)
< Content-Length: 1164
< Content-Type: application/ld+json
```

* Running `drupal router:rebuild` then `drupal cr all` resulted in same response :frowning_face: :


```
< HTTP/1.1 200 OK
< Date: Sat, 13 Jan 2018 21:16:54 GMT
* Server Apache/2.4.18 (Ubuntu) is not blacklisted
< Server: Apache/2.4.18 (Ubuntu)
< X-Powered-By: PHP/7.0.22-0ubuntu0.16.04.1
< Cache-Control: must-revalidate, no-cache, private
< Link: <http://localhost:8000/node/2>; rel="canonical"
< Link: <http://localhost:8000/node/2/delete>; rel="https://drupal.org/link-relations/delete-form"
< Link: <http://localhost:8000/node/2/edit>; rel="edit-form"
< Link: <http://localhost:8000/node/2/revisions>; rel="version-history"
< Link: <http://localhost:8000/node/2>; rel="https://drupal.org/link-relations/revision"
< Link: <http://localhost:8000/node?node=2>; rel="https://drupal.org/link-relations/create"
< Link: <http://localhost:8000/node/1>; rel="related"; title="Member Of"
< X-Drupal-Dynamic-Cache: MISS
< X-UA-Compatible: IE=edge
< Content-language: en
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< Expires: Sun, 19 Nov 1978 05:00:00 GMT
< X-Generator: Drupal 8 (https://www.drupal.org)
< Content-Length: 1164
< Content-Type: application/ld+json
```
I don't know what I'm doing that's different from @whikloj. I think we need to get a third person to test to break the tie.

**Comment by whikloj** [comment]
@mjordan have you tried uninstalling `islandora_collection`, `islandora_image` and `islandora`. Then pull @dannylamb's branch. Then reinstall all three modules? That is the only difference so far I can think of.

Not that we haven't had cases where things didn't appear right for me too. So it could be a ghost in the shell.

**Comment by mjordan** [comment]
Getting a WSOD while uninstalling the islandora module. Will continue....

**Comment by mjordan** [comment]
With a totally new and clean VM, using the GUI I uninstalled `islandora_image` (worked), `islandora_collection` (worked), but then when I try to uninstall `islandora`, I am told: `The website encountered an unexpected error. Please try again later.`

Not sure what to do next. Maybe see if anyone else can test?

**Comment by dannylamb** [comment]
@mjordan I have a gist that I use to get past this one: https://gist.github.com/dannylamb/0becb59b890403f8035915319bf086f1

If you get the mysql errors about moving databases, I brute force it with vim scripts to make an SQL script that drops all the offending tables.

**Comment by whikloj** [comment]
@dannylamb I actually stored that script in my claw-playbook directory so I can access it from inside the vagrant VM. Maybe we should distribute it with islandora until we can resolve that storage problem?

**Comment by Natkeeran** [comment]
@dannylamb 

Uninstalled islandora.  Applied the PR. Installed the modules. Created a collection item and an image item with a TN.  

Please see the curl response below.  the  rel="related" for collection membership exists.  I am not I am supposed to see one for the media bundle as well.  


```
root@clawtest:/var/www/html/drupal/web/modules/contrib/islandora# curl -v "http://165.227.35.195:8000/node/3?_format=jsonld"
*   Trying 165.227.35.195...
* Connected to 165.227.35.195 (165.227.35.195) port 8000 (#0)
> GET /node/3?_format=jsonld HTTP/1.1
> Host: 165.227.35.195:8000
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Tue, 30 Jan 2018 20:47:58 GMT
< Server: Apache/2.4.18 (Ubuntu)
< X-Powered-By: PHP/7.0.22-0ubuntu0.16.04.1
< Cache-Control: must-revalidate, no-cache, private
< Link: <http://165.227.35.195:8000/node/3>; rel="canonical"
< Link: <http://165.227.35.195:8000/node/3/delete>; rel="https://drupal.org/link-relations/delete-form"
< Link: <http://165.227.35.195:8000/node/3/edit>; rel="edit-form"
< Link: <http://165.227.35.195:8000/node/3/revisions>; rel="version-history"
< Link: <http://165.227.35.195:8000/node/3>; rel="https://drupal.org/link-relations/revision"
< Link: <http://165.227.35.195:8000/node?node=3>; rel="https://drupal.org/link-relations/create"
< Link: <http://165.227.35.195:8000/node/2>; rel="related"; title="Member Of"
< X-Drupal-Dynamic-Cache: MISS
< X-UA-Compatible: IE=edge
< Content-language: en
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< Expires: Sun, 19 Nov 1978 05:00:00 GMT
< X-Generator: Drupal 8 (https://www.drupal.org)
< X-Drupal-Cache: MISS
< Content-Length: 1204
< Content-Type: application/ld+json
< 
{"@graph":[{"@id":"http:\/\/165.227.35.195:8000\/node\/3?_format=jsonld","@type":["http:\/\/pcdm.org\/models#Object","http:\/\/schema.org\/ImageObject"],"http:\/\/purl.org\/dc\/terms\/title":[{"@value":"Test Image","@type":"http:\/\/www.w3.org\/2001\/XMLSchema#string"}],"http:\/\/schema.org\/author":[{"@id":"http:\/\/165.227.35.195:8000\/user\/1?_format=jsonld"}],"http:\/\/schema.org\/dateCreated":[{"@value":"2018-01-30T20:46:48+00:00","@type":"http:\/\/www.w3.org\/2001\/XMLSchema#dateTime"}],"http:\/\/schema.org\/dateModified":[{"@value":"2018-01-30T20:47:17+00:00","@type":"http:\/\/www.w3.org\/2001\/XMLSchema#dateTime"}],"http:\/\/pcdm.org\/models#memberOf":[{"@id":"http:\/\/165.227.35.195:8000\/node\/2?_format=jsonld"}],"http:\/\/pcdm.org\/models#hasFile":[{"@id":"http:\/\/165.227.35.195:8000\/media\/1?_format=jsonld"}]},{"@id":"http:\/\/165.227.35.195:8000\/user\/1?_format=jsonld","@type":["http:\/\/schema.org\/Person"]},{"@id":"http:\/\/165.227.35.195:8000\/node\/2?_format=jsonld","@type":["http:\/\/pcdm* Connection #0 to host 165.227.35.195 left intact
.org\/models#Collection","http:\/\/schema.org\/CollectionPage"]},{"@id":"http:\/\/165.227.35.195:8000\/media\/1?_format=jsonld","@type":["http:\/\/pcdm.org\/use#ThumbnailImage"]}]}root@clawtest:/var/www/html/drupal/web/modules/contrib/islandora# 
```

**Comment by dannylamb** [comment]
@Natkeeran You won't see media links if you're cURLing as anonymous. You need view permissions for an entity to see it as a link header. If you cURL with `-u admin:islandora` you should see them.

**Comment by Natkeeran** [comment]
@dannylamb 

related headers looks good.  There seems to be some inconsistency between edit and delete headers.  

```
curl -v -u admin:islandora "http://138.197.143.179:8000/node/2?_format=jsonld"
*   Trying 138.197.143.179...
* TCP_NODELAY set
* Connected to 138.197.143.179 (138.197.143.179) port 8000 (#0)
* Server auth using Basic with user 'admin'
> GET /node/2?_format=jsonld HTTP/1.1
> Host: 138.197.143.179:8000
> Authorization: Basic YWRtaW46aXNsYW5kb3Jh
> User-Agent: curl/7.55.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Wed, 31 Jan 2018 15:47:33 GMT
< Server: Apache/2.4.18 (Ubuntu)
< X-Powered-By: PHP/7.0.22-0ubuntu0.16.04.1
< Cache-Control: must-revalidate, no-cache, private
< Link: <http://138.197.143.179:8000/node/2>; rel="canonical"
< Link: <http://138.197.143.179:8000/node/2/delete>; rel="https://drupal.org/link-relations/delete-form"
< Link: <http://138.197.143.179:8000/node/2/edit>; rel="edit-form"
< Link: <http://138.197.143.179:8000/node/2/revisions>; rel="version-history"
< Link: <http://138.197.143.179:8000/node/2>; rel="https://drupal.org/link-relations/revision"
< Link: <http://138.197.143.179:8000/node?node=2>; rel="https://drupal.org/link-relations/create"
< Link: <http://138.197.143.179:8000/node/1>; rel="related"; title="Member Of"
< Link: <http://138.197.143.179:8000/media/1>; rel="related"; title="TN"
< X-Drupal-Dynamic-Cache: HIT
< X-UA-Compatible: IE=edge
< Content-language: en
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< Expires: Sun, 19 Nov 1978 05:00:00 GMT
< X-Generator: Drupal 8 (https://www.drupal.org)
< Content-Length: 1213
< Content-Type: application/ld+json
< 
{"@graph":[{"@id":"http:\/\/138.197.143.179:8000\/node\/2?_format=jsonld","@type":["http:\/\/pcdm.org\/models#Object","http:\/\/schema.org\/ImageObject"],"http:\/\/purl.org\/dc\/terms\/title":[{"@value":"test image 1","@type":"http:\/\/www.w3.org\/2001\/XMLSchema#string"}],"http:\/\/schema.org\/author":[{"@id":"http:\/\/138.197.143.179:8000\/user\/1?_format=jsonld"}],"http:\/\/schema.org\/dateCreated":[{"@value":"2018-01-31T15:43:57+00:00","@type":"http:\/\/www.w3.org\/2001\/XMLSchema#dateTime"}],"http:\/\/schema.org\/dateModified":[{"@value":"2018-01-31T15:47:04+00:00","@type":"http:\/\/www.w3.org\/2001\/XMLSchema#dateTime"}],"http:\/\/pcdm.org\/models#memberOf":[{"@id":"http:\/\/138.197.143.179:8000\/node\/1?_format=jsonld"}],"http:\/\/pcdm.org\/models#hasFile":[{"@id":"http:\/\/138.197.143.179:8000\/media\/1?_format=jsonld"}]},{"@id":"http:\/\/138.197.143.179:8000\/user\/1?_format=jsonld","@type":["http:\/\/schema.org\/Person"]},{"@id":"http:\/\/138.197.143.179:8000\/node\/1?_format=jsonld","@type":["http:* Connection #0 to host 138.197.143.179 left intact
\/\/pcdm.org\/models#Collection","http:\/\/schema.org\/CollectionPage"]},{"@id":"http:\/\/138.197.143.179:8000\/media\/1?_format=jsonld","@type":["http:\/\/pcdm.org\/use#ThumbnailImage"]}]}
```

**Comment by dannylamb** [comment]
@Natkeeran What's the inconsistency you're noticing with edit and delete links?

**Comment by Natkeeran** [comment]
@dannylamb 
```
< Link: <http://138.197.143.179:8000/node/2/delete>; rel="https://drupal.org/link-relations/delete-form"
< Link: <http://138.197.143.179:8000/node/2/edit>; rel="edit-form"
```
The delete link has the full url while the edit link just has the name/label/title?

**Comment by dannylamb** [comment]
@Natkeeran That's because edit-form is in the [IANA link registry](https://www.iana.org/assignments/link-relations/link-relations.xhtml).  It comes from https://tools.ietf.org/html/rfc6861.  Drupal had to mint their own relation for the delete-form, so they have to give the absolute url.  If they got it put in the registry we could just say 'delete-form'.

**Comment by Natkeeran** [comment]
@dannylamb 

Thank you for the clarification.  @whikloj if you are good, can we merge this?
